src/content/articles/why-typescript-doesnt-follow-strict-semantic-versioning.mdx

```diff
@@ -0,0 +1,278 @@
+A lockfile will ensure that installing dependencies in your project will always use the same versions.
+Using a lockfile is generally a good practice - especially for tools such as TypeScript that can substantially change behavior across versions.
+
+Note that when using a lockfile, you don't have to specify an exact _major.minor.patch_ version for `"typescript"` in your `package.json`.
+Your lockfile will specify an exact version of every dependency for you - generally defaulting to the latest stable version of any dependencies in your `package.json`, including `"devDependencies"`.
+
+### Intentional TypeScript Version Upgrades
+
+When you do want to update your TypeScript version, consider the following strategy:
+
+1. Create a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) to isolate just the changes for the new TypeScript version
+2. In the pull request body, link to the TypeScript release notes (e.g. [TypeScript 5.1](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-1.html)) and call out changes that are impactful to your project
+3. In the PR's files view, for each kind of change you had to make, add a comment to the first instance of the change explaining why it was necessary
+
+### Dealing with Breakages
+
+If and when a new TypeScript version causes a type error in your code, consider trying each of the following strategies in order until it is fixed:
+
+1. Fix small issues with the types to work with the new TypeScript behavior
+2. Refactor the types to reduce complexity and work with the new TypeScript behavior
+3. Use `any` on types you can't figure out, with a `// TODO` comment linked to a tracking issue/ticket for cleaning it up later
+4. Use `// @ts-expect-error` on remaining issues you can't figure out, with a `// TODO` comment linked to a tracking issue/ticket for cleaning it up later
+
+Let's look at those tips in more details.
+
+#### Straightforward, Clean Types
+
+TypeScript's type system is fantastically powerful and can represent some wild and wacky type operations.
+However, the more complex your types are, the more difficult they are to read, write, and update over time.
+TypeScript types follow simplicity principles from other types of code:
+
+- Prefer simple, easy-to-read code (types) whenever possible
+- Prefer well-named types that explain what they're doing
+- When code does become complex to the point of being difficult to read, consider using comments to explain the tricky details
+
+Simpler types are less likely to be broken by new TypeScript versions tinkering with nuances of the type system.
+They're also easier to debug when something becomes broken.
```

Comment:
This could be a chance to toot your horn, consider linking to a project (doesn't have to be yours either) that might exemplify these patterns.

Reply:
Ha, I think I'll pass - I have too much bashfulness and suspect others do too.